bpf/standalone/arc_prefetch.py

```diff
@@ -0,0 +1,416 @@
+    if (info == NULL || info->arc_read_ts == 0)
+        return 0;
+
+    bpf_trace_printk("arc_read_return %d %d\\n", info->zfs_read_ts,
+        info->arc_read_ts);
+
+    u64 elapsed = (bpf_ktime_get_ns() - info->arc_read_ts) / 1000;
+    lat_key lkey = {1, ARC_ISSUE_LATENCY, 0};
+    lkey.cpuid = bpf_get_smp_processor_id();
+    u32 slot = bpf_log2l(elapsed);
+    HIST_KEY_INITIALIZE(hist_lat_key, hkey, lkey, slot);
+    read_latency.increment(hkey);
+
+    average_t *average = read_average.lookup(&lkey);
+    if (average == NULL) {
+         average_t initial_average = {1, elapsed};
+         read_average.update(&lkey, &initial_average);
+         return 0;
+    }
+    average->count += 1;
+    average->sum += elapsed;
+    return 0;
+}
+
+int zio_read_entry(struct pt_regs *ctx, zio_t *zio)
+{
+    u32 tid = bpf_get_current_pid_tgid();
+    arc_prefetch_info_t *info = arc_prefetch_info.lookup(&tid);
+    if (info == NULL || info->arc_read_ts == 0)
+        return 0;
+    bpf_trace_printk("zio_read_entry %d %d\\n", info->zfs_read_ts,
+         info->arc_read_ts);
+
+    info->zio = zio;
+    return 0;
+}
+
+int zio_read_return(struct pt_regs *ctx)
+{
+    u32 tid = bpf_get_current_pid_tgid();
+    u64 zio_exit_ts = bpf_ktime_get_ns();
+    arc_prefetch_info_t *info = arc_prefetch_info.lookup(&tid);
+    if (info == NULL || info->zio == NULL)
+        return 0;
+    bpf_trace_printk("zio_read_return %d %d\\n",
+        info->zfs_read_ts, info->arc_read_ts);
+
+    zio_read_exit_time.update(&info->zio, &zio_exit_ts);
+    return 0;
+}
+
+int arc_read_done_entry(struct pt_regs *ctx, zio_t *zio)
+{
+    arc_buf_hdr_t *hdr = (arc_buf_hdr_t *) zio->io_private;
+    u64 zero = 0;
+
+    u64 *zio_exit_ts = zio_read_exit_time.lookup(&zio);
+    if (zio_exit_ts == NULL || *zio_exit_ts == 0) {
+            return 0;
+    }
+
+    u64 zio_ts = *zio_exit_ts;
+    u64 elapsed = (bpf_ktime_get_ns() - zio_ts) / 1000;
+    bpf_trace_printk("zio_read_return %d \\n", zio_ts, elapsed);
+
+    hist_lat_key hkey = { };
+    lat_key *lkey = HIST_KEY_GET_AGGKEY(&hkey);
+    u32 count_index;
+    if (hdr->b_flags & ARC_FLAG_PREFETCH) {
+        count_index = ARC_PREFETCH_ZIO_COUNT;
+        lkey->index = ARC_PREFETCH_ZIO_LATENCY;
+    } else {   // normal zio
+        count_index = ARC_NORMAL_ZIO_COUNT;
+        lkey->index = ARC_NORMAL_ZIO_LATENCY;
+    }
+    lkey->t = 1;
+    lkey->cpuid = bpf_get_smp_processor_id();
+
+    HIST_KEY_SET_SLOT(&hkey, bpf_log2(elapsed));
+    arc_count.increment(count_index);
+    read_latency.increment(hkey);
+    zio_read_exit_time.update(&zio, &zero);
+    average_t *average = read_average.lookup(lkey);
+    if (average == NULL) {
+         average_t initial_average = {1, elapsed};
+         read_average.update(lkey, &initial_average);
+         return 0;
+    }
+    average->count += 1;
+    average->sum += elapsed;
+
+    return 0;
+ }
+
+"""
+
+
+class BCCMapIndex(Enum):
```

Comment:
Sounds like this could be moved to common - say `lib/bcchelper.py` ? and also a common method `getCDefinitions()` ?